### PR TITLE
8268139: CDS ArchiveBuilder may reference unloaded classes

### PR DESCRIPTION
**Pull request by iklam**
During CDS dump, we enter a safepoint (e.g., `VM_PopulateDynamicDumpSharedSpace`) and run `ArchiveBuilder::gather_source_objs()` to collect all classes to be dumped. However, some of these classes may belong to custom class loaders that are no longer alive. While we are still inside the safepoint, concurrent GC operations may free these class loaders (see https://github.com/openjdk/jdk/pull/4286). Thus `ArchiveBuilder` may end up referencing unloaded classes.

The fix:

(1) Add a class `k` to ArchiveBuilder only if `k->class_loader_data()->is_alive()`

(2) All the classes added to `ArchiveBuilder` are taken from `DumpTimeSharedClassTable`. When a class is loaded, it's added to this table. When a class is unloaded, it's removed from this table (via `SystemDictionaryShared::remove_dumptime_info`). This table is protected by `DumpTimeTable_lock`. To make sure that this table is not changed during the CDS safepoint, we grab this lock before calling `ArchiveBuilder::gather_source_objs()`, and release this lock only when the CDS dump is finished. As a result, in step (1) above, when we call `k->class_loader_data()->is_alive()`, we are sure that `k` itself has not been deallocated and is still a valid pointer.

Testing -- Mach5 tiers 1-4 in progress

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8268139](https://bugs.openjdk.java.net/browse/JDK-8268139): CDS ArchiveBuilder may reference unloaded classes

### Reviewers
 * [Coleen Phillimore](https://openjdk.java.net/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to 49ae57aed856f445bc3e0563123c6eacc199e32c
 * [Calvin Cheung](https://openjdk.java.net/census#ccheung) (@calvinccheung - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4322/head:pull/4322` \
`$ git checkout pull/4322`

Update a local copy of the PR: \
`$ git checkout pull/4322` \
`$ git pull https://git.openjdk.java.net/jdk pull/4322/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4322`

View PR using the GUI difftool: \
`$ git pr show -t 4322`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4322.diff">https://git.openjdk.java.net/jdk/pull/4322.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/classfile/systemDictionaryShared.cpp [hunk 1]
```diff
@@ -1499,13 +1507,18 @@ bool SystemDictionaryShared::has_class_failed_verification(InstanceKlass* ik) {
   return (p == NULL) ? false : p->failed_verification();
 }
 
+static bool is_class_alive(InstanceKlass* k) {
+  assert_lock_strong(DumpTimeTable_lock);
+  return k->class_loader_data()->is_alive();
+}
+
```
Comment (coleenp) [inline src/hotspot/share/classfile/systemDictionaryShared.cpp:1514]: There's a k->is_loader_alive() call, so you don't need this.

**Review by coleenp — APPROVED** [review]
Except a minor comment, this looks good.

**Review by calvinccheung — APPROVED** [review]
Looks good. In systemDictionaryShared.cpp

 `432   assert(DumpTimeTable_lock->owned_by_self(), "sanity");`

Can the above be changed to `assert_lock_strong(DumpTimeTable_lock);`?

**Comment by iklam** [comment]
> Looks good. In systemDictionaryShared.cpp
> 
> `432 assert(DumpTimeTable_lock->owned_by_self(), "sanity");`
> 
> Can the above be changed to `assert_lock_strong(DumpTimeTable_lock);`?

Hi Calvin, thanks for the review. I've changed it as you suggested. This makes the code more consistent.

**Review by calvinccheung — APPROVED** [review]

**Comment by iklam** [comment]
Thanks @coleenp and @calvinccheung for the review
/integrate

**Pull request closed without merging**